### PR TITLE
Detection for nested folders (CLI)

### PR DESCRIPTION
**Pull request by sizov-kirill**
### Summary
1. This PR repeated functionality that was implemented in #623, but now recursive search is working in CLI command as well.

2. Changed strategy for recursive format detection: the function `env.detect_dataset` now returns a list of detected formats at all recursion levels (in case when each step of recursion returned a more than one format).

3. Fixed detection for LFW format.

4. Added detection for Cityscapes format.

5. Open Images: allowed to store annotations file in root path as well

### Checklist

- [ ] I submit my changes into the `develop` branch
- [ ] I have added description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md)
- [ ] I have updated the [documentation](
  https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly
- [ ] I have added tests to cover my changes
- [ ] I have [linked related issues](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

### License

- [ ] I submit _my code changes_ under the same [MIT License](
  https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: tests/cli/test_detect_format.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def test_ambiguous(self):
             output_file = io.StringIO()
 
             with contextlib.redirect_stdout(output_file):
-                run(self, 'detect-format', test_dir)
+                run(self, 'detect-format', '--depth', '0', test_dir)
```
Comment (zhiltsov-max) [inline tests/cli/test_detect_format.py]: This test doesn't seem valid now, because it should be ambiguous either way. Could you check it?

Comment (zhiltsov-max) [inline tests/cli/test_detect_format.py:51]: This test doesn't seem valid now, because it should be ambiguous either way. Could you check it?

**Comment by zhiltsov-max** [comment]
Please fix conflicts.

File: datumaro/components/environment.py [hunk 2]
```diff
@@ -240,11 +240,27 @@ def make_transform(self, name, *args, **kwargs):
...
 
-    def detect_dataset(self, path):
-        return detect_dataset_format(
-            (
-                (format_name, importer.detect)
-                for format_name, importer in self.importers.items.items()
-            ),
-            path,
-        )
+    def detect_dataset(self, path, rejection_callback=None, depth=1):
```
Comment (zhiltsov-max) [inline datumaro/components/environment.py]:
```suggestion
    def detect_dataset(self, path: str, rejection_callback: Optional[RejectionCallback] = None, depth: int = 1) -> List[str]:
```

`RejectionCallback = Callable[[str, RejectionReason, str], None]`

**Comment by zhiltsov-max** [comment]
Please update changelog

File: datumaro/plugins/open_images_format.py [hunk 3]
```diff
@@ -175,7 +175,10 @@ def __init__(self, path, image_meta=None):
 
         self._dataset_dir = path
 
-        self._annotation_files = os.listdir(osp.join(path, OpenImagesPath.ANNOTATIONS_DIR))
+        self._annotation_dir = osp.join(path, OpenImagesPath.ANNOTATIONS_DIR)
```
Comment (zhiltsov-max) [inline datumaro/plugins/open_images_format.py:178]: Please decide if you want to support different layouts and document it, if so.

Comment (sizov-kirill) [reply datumaro/plugins/open_images_format.py:178]: [Decision already applied](https://github.com/openvinotoolkit/datumaro/blob/7676f2e42ce2e65351c7a2b15eecc0ec0aeecb90/datumaro/plugins/open_images_format.py#L178-L180): if `annotations` directory exists then search annotation files in it, else search annotation file in root of input path.

**Pull request merged**